### ISSUE TITLE
.blob() support

### ISSUE THREAD

**Issue by jonathan-fielding** [bitinn/node-fetch#78]
Hi

So I have been using node-fetch for a while now and for the first time I needed to request an image and get its size, now in the client I can do.

```
fetch('imageURL').then(function(response) {
    return response.blob();
}).then(function(resp){
    console.log(resp.size)
});
```

But unfortunately blob isn't supported by node-fetch so this won't work

Thanks

Jonathan

**Comment by bitinn** [comment]
Yeah, though you should be able to use the `res.body` stream instead.

with modules like https://www.npmjs.com/package/stream-meter (it's a simple enough task you can roll your own module).

**Comment by bogas04** [comment]
@bitinn first of all thanks for this module, a fresh breath over `request`.

Can you elaborate how can we convert response to a blob ?

**Comment by bitinn** [comment]
@bogas04 `blob` is a very client-side thing, node.js prefer stream and buffers when handling data.

see https://github.com/substack/stream-handbook for some explanation.

**Comment by bogas04** [comment]
@bitinn Thanks for the link! Will have a look.

**Comment by jonathan-fielding** [comment]
@bitinn the problem comes if I want to write JavaScript that can run either clientside or serverside, it makes sense in that use case to support blob,

**Comment by bogas04** [comment]
@jonathan-fielding You shouldn't be using `node-fetch` for isomorphic js AFAIK. [This](https://github.com/matthew-andrews/isomorphic-fetch) might help.

Anyway, you can always check where your JS code is running in (React has ExecutionEnvironment.canUseDOM or something like that) and change your code accordingly.

**Comment by jonathan-fielding** [comment]
@bogas04, thanks for your reply 😀, 

node-fetch is actually used by that isomorphic-fetch you linked to, if its node it uses node-fetch, if its browser it uses the browser fetch polyfill.

Sure I can check the environment the code is running in but the benefit you get from writing Universal JavaScript is that it is the same code that works on both clientside and serverside.

**Comment by bogas04** [comment]
@jonathan-fielding Even then, in theory, I think node-fetch itself shouldn't get bloated only for isomorphic feels.

In your case you can probably create a module or something.

A naive untested implementation just to get you started:

```js
import toBlob from 'isomorphic-fetch-blob';

fetch(url).then(toBlob).then(image => doStuff(image));
// this differs from the pure isomorphic form only in the syntax to convert to blob.
fetch(url).then(r => r.blob()).then(image => doStuff(image));
```

```js
// isomorphic-fetch-blob
import { Writable } from 'stream';

const isBrowser = () => typeof window !== 'undefined';

let buffer;
const ws = Writable();
ws._write = (chunk, enc, next) => {
    buffer += chunk;
    next();
};

export default (data) => {
  if(isBrowser()) {
    return data.blob();
  } else {
    return new Promise((resolve, reject) => data.body
      .pipe(ws).on('end', () => resolve(buffer))
    );
  }
};
```

**Comment by bitinn** [comment]
@jonathan-fielding I have argued against isomorphism in this specific case, ideally you shouldn't have to worry about environment, but the tradeoff is to implement blob on node.js, when buffer and stream are the default choice (superior choice if you think about how stream works).

I have not seen a convincing reason to use blob on node.js, there is no native api that use blob, unlike browser, you end up needing to convert it back to buffer...

**Comment by bogas04** [comment]
@bitinn yeah I'm with you on this, the core principle of NodeJS was to avoid buffering pretty much always, ignoring that and favoring cleaner code/ easy maintenance isn't really ringing bells for me either.

**Comment by kirill-konshin** [comment]
We should probably add some non standard methods like `.buffer()` to get the body as a `Buffer` instance...

**Comment by bitinn** [comment]
FYI, if you need to buffer a small response on Node.js, you can now use `res.buffer()` since v1.6.0; for anything we still strongly recommend using `res.body` as stream. That's the proper and performant way on Node.js.

**Comment by TimothyGu** [comment]
I'm looking at this issue right now. The thing about Blob is that it doesn't have any outlets in Node.js -- once you pass buffers into the Blob, you will not be able to access them any more, at least not with standard-compliant methods. So @jonathan-fielding, what is your use case?

**Comment by jonathan-fielding** [comment]
The use case I had was I was trying to assert some information about some images such a file size and type.

**Comment by jimmywarting** [comment]
This could work in both node-fetch and browsers fetch

```js
res = await fetch(...)
buffer = await res.arrayBuffer()
console.log(buffer.byteLength) // the final size
```
Apart from async/await but convert it to `.then` and it will be fine

**Comment by TimothyGu** [comment]
Fixed in #200.
